### PR TITLE
1609 Submission History

### LINKED ISSUE #1609: As a TDP user I want a list of prior submissions for the selected quarter/STT
**Description:**
_Provide a brief background and justification for this issue_
This is for release 2 (pre parsing submission history), would be used by both Grantee Data Analysts and OFA users (anyone who can see the data files tab).
As a TDP user I want to see a list of prior submissions for the selected quarter/STT.

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [ ] Submission history tab displays a table of files for all sections matching Figma specs
- [ ] [USWDS Segmented button group](https://designsystem.digital.gov/components/button-group/#segmented-button-group) is utilized.
- [ ] Section title ("Section [x] - [Section name]" should be implemented with a custom styled table caption as mentioned in Figma implementation notes
- [ ] Filename, timestamp, author are visible as columns.
- [ ] Columns have alternating background colors to break these rows up visually
- [ ] Tables are paginated for submissions lists over 5 records
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [ ] Implement button group
- [ ] Create new submission history tables
- [ ] Add pagination to tables
- [ ] Implement alternating background colors for rows
- [ ] Populate tables with relevant file metadata
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 

**Supporting Documentation:**
image
- [Figma link](https://www.figma.com/file/irgQPLTrajxCXNiYBTEnMV/TDP-Mockups-For-Feedback?t=a4NHkaguYUIPL7UB-1) - 
- [Segmented button group](https://designsystem.digital.gov/components/button-group/#segmented-button-group)
- [Pagination component](https://designsystem.digital.gov/components/pagination/)
- [HTML prototype for v3.x (parsing dependent)](https://reitermb.github.io/HTML-Table-example/)

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_ 

- Do we want to display an "author" or "submitted by" field?
   - Yes
- As we move forward what will regional staff users have access to (if anything) under "Data Files". Part of the ongoing discussion of research insight around the usefulness of Grantees/Federal staff having the same view of the same data.
- Might there be a need in submission volume for API/backend work to support pagination?

### PR DESCRIPTION
**Pull request by jtimpe**
## Summary of Changes
Pull request closes #1609

- Introduces a Current Submission/Submission History tab on the Data Files page
- Submission History tab shows a complete list of submitted data for the selected STT/year/quarter search combo, paginated for each section
- Added generic Segmented Control and Pagination components in addition to the new tab content  

This also includes changes from #1107

- Adds s3 object versioning to data files, updates the `download` view to handle a new version id

## How to Test

```bash
cd tdrs-frontend && npm run start

# in another terminal
cd tdrs-backend && docker-compose up
```

1. Open http://localhost:3000/ and sign in.
2. Navigate to Data Files and search for a year/quarter. The old functionality on the Current Submission tab should work the same as before.
3. Click into the Submission History tab. If there are results for a year/quarter, they should be displayed along with the submission date and a link to download the file.

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] Implement button group (SegmentedControl)
+ [x] Create new Submission History tables
   - Alternating background color for rows
   - Populated with Submitted On, Submitted By, and File Name metadata
   - File Name should be clickable to download the file
+ [x] Add pagination to tables
   - Page size of 5
   - Not shown when tables are empty, or only one page of results is shown
+ [ ] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/gunicorn_start.sh [hunk 1]
```diff
@@ -26,7 +26,7 @@ sleep 5
 celery -A tdpservice.settings beat -l info --scheduler django_celery_beat.schedulers:DatabaseScheduler &
 
 # rebuild and repopulate elastic search indexes
-python manage.py search_index --rebuild -f
+# python manage.py search_index --rebuild -f
```
Comment (andrew-jameson) [inline tdrs-backend/gunicorn_start.sh]: Should this be commented out?

File: tdrs-backend/tdpservice/data_files/models.py [hunk 2]
```diff
@@ -173,6 +174,11 @@ def fiscal_year(self):
 
         return f"{self.year} - {self.quarter} {quarter_month_str}"
 
+    @property
+    def submitted_by(self):
+        """Return the correct filename for this data file."""
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/data_files/models.py]:
```suggestion
        """Return the author as a string for this data file."""
```

File: tdrs-backend/tdpservice/data_files/views.py [hunk 3]
```diff
@@ -158,6 +168,44 @@ def download(self, request, pk=None):
...
+        results = DataFile.objects.order_by('section', '-version').distinct('section')
+
+        for result in results:
+            # try:
+            #     data_file = DataFile.objects.get(
+            #         stt=stt, quarter=quarter, year=year, section=section)
+
+            # except DataFile.DoesNotExist:
+            #     latest_submissions[section] = None
+
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/data_files/views.py]:
```suggestion
```

Should we delete this old code?

File: tdrs-backend/tdpservice/data_files/views.py [hunk 4]
```diff
@@ -158,6 +168,44 @@ def download(self, request, pk=None):
...
+            latest_submissions[result.section] = self.serializer_class(result).data
+
+        return Response(
+            {'results': latest_submissions},
+            200
+        )
+
+        # return self.list(self, request)
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/data_files/views.py]:
```suggestion
```

L202-L204 Makes this obsolete, yes?

File: tdrs-backend/tdpservice/data_files/views.py [hunk 5]
```diff
@@ -139,6 +144,11 @@ def download(self, request, pk=None):
         """Retrieve a file from s3 then stream it to the client."""
         record = self.get_object()
 
+        # response = FileResponse(
+        #     FileWrapper(record.file),
+        #     filename=record.original_filename
+        # )
+        # return response
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/data_files/views.py]:
```suggestion
```

File: tdrs-backend/gunicorn_start.sh (hunk 1 above)
Comment (jtimpe) [reply tdrs-backend/gunicorn_start.sh]: yeah, this was causing some issues with the database migrations when there was already data in the deployed environment. it's not really needed as elastic will auto-migrate indexed data (safely) on deployment anyway.

Comment (jtimpe) [reply tdrs-backend/gunicorn_start.sh]: removed it altogether

File: tdrs-backend/tdpservice/data_files/views.py [hunk 6]
```diff
@@ -98,16 +102,17 @@ def create(self, request, *args, **kwargs):
...
         bucket_name = settings.AWS_S3_DATAFILES_BUCKET_NAME
-        versions = s3.client.list_object_versions(Bucket=bucket_name)
+        versions = s3.client.list_object_versions(Bucket=bucket_name, Prefix=prefix)
         for version in versions['Versions']:
             file_path = version['Key']
             if file_name in file_path:
                 if version['IsLatest']:
-                    return (version['VersionId'])
+                    if version['VersionId'] != 'null':
+                        return (version['VersionId'])
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/data_files/views.py]: Worth `and`-ing the conditionals b/c there are no elses? I don't think it's worth code-golfing via lambda/list comprehensions.

File: tdrs-backend/gunicorn_start.sh [hunk 7]
```diff
@@ -25,9 +25,6 @@ sleep 5
 # celery -A tdpservice.settings --broker=$REDIS_URI flower &
 celery -A tdpservice.settings beat -l info --scheduler django_celery_beat.schedulers:DatabaseScheduler &
 
-# rebuild and repopulate elastic search indexes
-python manage.py search_index --rebuild -f
```
Comment (jtimpe) [inline tdrs-backend/gunicorn_start.sh:29]: removed this as it was causing the following issue when deploying

```
2023-01-25T09:17:24.24-0800 [APP/PROC/WEB/0] ERR [2023-01-25 17:17:24,248: INFO/MainProcess] DatabaseScheduler: Schedule changed.
   2023-01-25T09:17:24.31-0800 [APP/PROC/WEB/0] OUT Creating index 't1_submissions'
   2023-01-25T09:17:24.41-0800 [APP/PROC/WEB/0] OUT Creating index 't2_submissions'
   2023-01-25T09:17:24.51-0800 [APP/PROC/WEB/0] OUT Creating index 't4_submissions'
   2023-01-25T09:17:24.61-0800 [APP/PROC/WEB/0] OUT Creating index 't5_submissions'
   2023-01-25T09:17:24.73-0800 [APP/PROC/WEB/0] OUT Creating index 't6_submissions'
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] OUT Indexing 3288 'T1' objects
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR Traceback (most recent call last):
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django/db/backends/utils.py", line 84, in _execute
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR return self.cursor.execute(sql, params)
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR psycopg2.errors.UndefinedColumn: column search_indexes_t1.record does not exist
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR LINE 1: ...OR WITH HOLD FOR SELECT "search_indexes_t1"."id", "search_in...
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR ^
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR The above exception was the direct cause of the following exception:
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR Traceback (most recent call last):
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django/db/models/sql/compiler.py", line 1175, in execute_sql
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR cursor.execute(sql, params)
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django/db/backends/utils.py", line 66, in execute
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR return executor(sql, params, many, context)
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django/db/backends/utils.py", line 79, in _execute
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR with self.db.wrap_database_errors:
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django/db/utils.py", line 90, in __exit__
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR raise dj_exc_value.with_traceback(traceback) from exc_value
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django/db/backends/utils.py", line 84, in _execute
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR return self.cursor.execute(sql, params)
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR django.db.utils.ProgrammingError: column search_indexes_t1.record does not exist
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR LINE 1: ...OR WITH HOLD FOR SELECT "search_indexes_t1"."id", "search_in...
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR ^
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR During handling of the above exception, another exception occurred:
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR Traceback (most recent call last):
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/app/manage.py", line 31, in <module>
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR main()
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/app/manage.py", line 27, in main
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR execute_from_command_line(sys.argv)
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR utility.execute()
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django/core/management/__init__.py", line 413, in execute
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR self.fetch_command(subcommand).run_from_argv(self.argv)
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django/core/management/base.py", line 354, in run_from_argv
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR self.execute(*args, **cmd_options)
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django/core/management/base.py", line 398, in execute
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR output = self.handle(*args, **options)
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django_elasticsearch_dsl/management/commands/search_index.py", line 159, in handle
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR self._rebuild(models, options)
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django_elasticsearch_dsl/management/commands/search_index.py", line 140, in _rebuild
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR self._populate(models, options)
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django_elasticsearch_dsl/management/commands/search_index.py", line 117, in _populate
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR doc().update(qs, parallel=parallel)
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django_elasticsearch_dsl/documents.py", line 195, in update
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR return self._bulk(
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django_elasticsearch_dsl/documents.py", line 179, in _bulk
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR return self.bulk(*args, **kwargs)
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django_elasticsearch_dsl/documents.py", line 146, in bulk
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR return bulk(client=self._get_connection(), actions=actions, **kwargs)
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/elasticsearch/helpers/actions.py", line 410, in bulk
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR for ok, item in streaming_bulk(
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/elasticsearch/helpers/actions.py", line 319, in streaming_bulk
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR for bulk_data, bulk_actions in _chunk_actions(
   2023-01-25T09:17:24.91-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/elasticsearch/helpers/actions.py", line 155, in _chunk_actions
   2023-01-25T09:17:24.92-0800 [APP/PROC/WEB/0] ERR for action, data in actions:
   2023-01-25T09:17:24.92-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django_elasticsearch_dsl/documents.py", line 170, in _get_actions
   2023-01-25T09:17:24.92-0800 [APP/PROC/WEB/0] ERR for object_instance in object_list:
   2023-01-25T09:17:24.92-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django/db/models/query.py", line 353, in _iterator
   2023-01-25T09:17:24.92-0800 [APP/PROC/WEB/0] ERR yield from self._iterable_class(self, chunked_fetch=use_chunked_fetch, chunk_size=chunk_size)
   2023-01-25T09:17:24.92-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django/db/models/query.py", line 51, in __iter__
   2023-01-25T09:17:24.92-0800 [APP/PROC/WEB/0] ERR results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
   2023-01-25T09:17:24.92-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/django/db/models/sql/compiler.py", line 1178, in execute_sql
   2023-01-25T09:17:24.92-0800 [APP/PROC/WEB/0] ERR cursor.close()
   2023-01-25T09:17:24.92-0800 [APP/PROC/WEB/0] ERR psycopg2.errors.InvalidCursorName: cursor "_django_curs_140039043520320_sync_1" does not exist
   2023-01-25T09:17:29.09-0800 [APP/PROC/WEB/0] ERR [2023-01-25 17:17:29,095: INFO/MainProcess] DatabaseScheduler: Schedule changed.
   2023-01-25T09:17:35.78-0800 [APP/PROC/WEB/0] ERR [2023-01-25 17:17:35 +0000] [726] [DEBUG] Closing connection.
   2023-01-25T09:18:01.11-0800 [HEALTH/0] ERR Timed out after 1m0s (30 attempts) waiting for readiness check to succeed: failed to make TCP connection to 10.255.59.197:8080: dial tcp 10.255.59.197:8080: connect: connection refused
   2023-01-25T09:18:01.11-0800 [CELL/0] ERR Failed after 1m0.388s: readiness health check never passed.
   2023-01-25T09:18:01.11-0800 [CELL/SSHD/0] OUT Exit status 0
   2023-01-25T09:18:05.81-0800 [APP/PROC/WEB/0] ERR [2023-01-25 17:18:05 +0000] [726] [DEBUG] Closing connection.
```

I don't believe this line is needed: `django-elasticsearch-dsl` will manage its own migrations and re-indexing anyway, there's no point in forcing it after deployments. I don't fully understand the root cause, but I believe this was a fact of calling the reindex function at the wrong time when database migrations hadn't fully taken effect.

Comment (n0remac) [reply tdrs-backend/gunicorn_start.sh:29]: Sorry about that. I thought we had removed that line already.

**Pull request merged**